### PR TITLE
Ra 536 prototype

### PR DESCRIPTION
**Pull request by ingeknudsen**

### CODE REVIEW

File: radix-operator/job/handler.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+			return nil
+		}
+
+		return err
+	}
+
+	syncJob := radixJob.DeepCopy()
+	logger.Infof("Sync job %s", syncJob.Name)
+
+	radixRegistration, err := t.radixclient.RadixV1().RadixRegistrations().Get(syncJob.Spec.AppName, metav1.GetOptions{})
```
Comment (keaaa) [inline radix-operator/job/handler.go]: as agreed refactor away the need for radixRegistration by moving cloneUrl into job struct

File: charts/radix-operator/templates/radix-apps-rbac.yaml [hunk 2]
```diff
@@ -40,6 +40,16 @@ rules:
   - watch
   - patch
   - create
+- apiGroups:
+  - radix.equinor.com
```
[the remark this answers is not in the document]
Comment (ingeknudsen) [reply charts/radix-operator/templates/radix-apps-rbac.yaml]: Done

File: charts/radix-operator/templates/radix-pipeline-rbac.yaml [hunk 3]
```diff
@@ -14,6 +14,7 @@ rules:
   - radix.equinor.com
   resources:
   - radixapplications
+  - radixjobs
```
[the remark this answers is not in the document]
Comment (ingeknudsen) [reply charts/radix-operator/templates/radix-pipeline-rbac.yaml]: Done

File: pkg/apis/radix/v1/radixjobtypes.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+//RadixJobList is a list of Radix jobs
+type RadixJobList struct {
+	meta_v1.TypeMeta `json:",inline" yaml:",inline"`
+	meta_v1.ListMeta `json:"metadata" yaml:"metadata"`
+	Items            []RadixJob `json:"items" yaml:"items"`
+}
+
+//RadixJobStep holds status for a single step
+type RadixJobStep struct {
+	Name      string            `json:"name" yaml:"name"`
```
[the remark this answers is not in the document]
Comment (ingeknudsen) [reply pkg/apis/radix/v1/radixjobtypes.go]: Name could be build-frontend, so it is open ended

File: pkg/apis/radix/v1/radixjobtypes.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+	TargetEnvs []string          `json:"targetEnvironments" yaml:"targetEnvironments"`
+	Steps      []RadixJobStep    `json:"steps" yaml:"steps"`
+}
+
+// RadixJobCondition Holds the condition of a job
+type RadixJobCondition string
+
+// These are valid conditions of a deployment.
+const (
+	JobQueued    RadixJobCondition = "Queued"
```
[the remark this answers is not in the document]
Comment (ingeknudsen) [reply pkg/apis/radix/v1/radixjobtypes.go]: Done

File: pkg/apis/pipeline/pipeline.go [hunk 6]
```diff
@@ -1,41 +1,34 @@
...
+		Definition{v1.Build, []StepType{ApplyConfigStep, BuildStep}},
+		Definition{v1.Promote, []StepType{PromoteStep}}}
 }
 
 // GetPipelineFromName Gets pipeline from string
 func GetPipelineFromName(name string) (*Definition, error) {
 	// Default to build-deploy for backward compatibility
 	if name == "" {
-		return &Definition{BuildDeploy, []StepType{ApplyConfigStep, BuildStep, DeployStep}}, nil
+		return &Definition{v1.BuildDeploy, []StepType{ApplyConfigStep, BuildStep, DeployStep}}, nil
```
[the remark this answers is not in the document]
Comment (ingeknudsen) [reply pkg/apis/pipeline/pipeline.go]: Done

File: pkg/apis/job/job.go [hunk 7]
```diff
@@ -0,0 +1,492 @@
...
+
+func (job *Job) syncStatuses() (stopReconciliation bool, err error) {
+	stopReconciliation = false
+
+	allJobs, err := job.radixclient.RadixV1().RadixJobs(job.radixJob.Namespace).List(metav1.ListOptions{})
+	if err != nil {
+		err = fmt.Errorf("Failed to get all RadixJobs. Error was %v", err)
+	}
+
+	if job.radixJob.Spec.Stop {
```
[the remark this answers is not in the document]
Comment (ingeknudsen) [reply pkg/apis/job/job.go]: Done

File: pkg/apis/job/job.go [hunk 8]
```diff
@@ -0,0 +1,492 @@
...
+			return false, err
+		}
+
+		return true, nil
+	}
+
+	return
+}
+
+func (job *Job) noOtherRunningJobOnBranch(allJobs []v1.RadixJob) bool {
```
[the remark this answers is not in the document]
Comment (ingeknudsen) [reply pkg/apis/job/job.go]: Done

File: pkg/apis/job/job.go [hunk 9]
```diff
@@ -0,0 +1,492 @@
...
+	return !isOtherJobRunningOnBranch(job.radixJob, allJobs)
+}
+
+func isOtherJobRunningOnBranch(job *v1.RadixJob, allJobs []v1.RadixJob) bool {
+	for _, rj := range allJobs {
+		if rj.GetName() != job.GetName() &&
+			(job.Spec.Build.Branch != "" && job.Spec.Build.Branch == rj.Spec.Build.Branch) &&
```
[the remark this answers is not in the document]
Comment (ingeknudsen) [reply pkg/apis/job/job.go]: Done

File: radix-operator/job/handler.go (hunk 1 above)
Comment (ingeknudsen) [reply radix-operator/job/handler.go]: Done

File: pkg/apis/job/job.go [hunk 10]
```diff
@@ -0,0 +1,492 @@
...
+		return err
+	}
+
+	jobStatus := getJobConditionFromJobStatus(pipelinejob.Status)
+
+	job.radixJob.Status.Condition = jobStatus
+	job.radixJob.Status.Started = pipelinejob.Status.StartTime
+
+	if len(pipelinejob.Status.Conditions) > 0 {
+		job.radixJob.Status.Ended = &pipelinejob.Status.Conditions[0].LastTransitionTime
```
[the remark this answers is not in the document]
Comment (ingeknudsen) [reply pkg/apis/job/job.go]: Done

File: pkg/apis/job/kubejob.go [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+	if pushImage {
+		return "1"
+	}
+
+	return "0"
+}
+
+func getJobConditionFromJobStatus(jobStatus batchv1.JobStatus) v1.RadixJobCondition {
+	var status v1.RadixJobCondition
+	if jobStatus.Active > 0 {
```
[the remark this answers is not in the document]
Comment (ingeknudsen) [reply pkg/apis/job/kubejob.go]: Done

File: pkg/apis/job/job.go [hunk 12]
```diff
@@ -0,0 +1,492 @@
...
+		}
+	}
+}
+
+func (job *Job) syncStatuses() (stopReconciliation bool, err error) {
+	stopReconciliation = false
+
+	allJobs, err := job.radixclient.RadixV1().RadixJobs(job.radixJob.Namespace).List(metav1.ListOptions{})
+	if err != nil {
+		err = fmt.Errorf("Failed to get all RadixJobs. Error was %v", err)
```
[the remark this answers is not in the document]
Comment (ingeknudsen) [reply pkg/apis/job/job.go]: Done

File: pkg/apis/job/job.go [hunk 13]
```diff
@@ -0,0 +1,492 @@
...
+	if restoredStatus, ok := job.radixJob.Annotations[kube.RestoredStatusAnnotation]; ok {
+		if job.radixJob.Status.Condition == "" {
+			var status v1.RadixJobStatus
+			err := json.Unmarshal([]byte(restoredStatus), &status)
+			if err != nil {
+				log.Error("Unable to get status from annotation", err)
+				return
+			}
+
+			job.radixJob.Status.Condition = status.Condition
```
[the remark this answers is not in the document]
Comment (ingeknudsen) [reply pkg/apis/job/job.go]: Done

**Pull request merged**